File: NodeDesignPatterns/builder.js
"use strict";
let doc = `

              +----------+
              |  Client  |
              +----------+
                    |
                    | uses
                    v
          +-------------------+
          |   UserBuilder     |
          +-------------------+
          | set_name()        |
          | set_email()       |
          | set_age()         |
          | build()           |
          +-------------------+
                    |
                    | creates
                    v
              +----------+
              |   User   |
              +----------+
              | name     |
              | email    |
              | age      |
              +----------+


`;
console.log(doc);
class User {
    name;
    age;
    email;
    address;
    dob;
    constructor(name, age, email, address, dob) {
        this.name = name;
        this.age = age;
        this.email = email;
        this.address = address;
        this.dob = dob;
        this.name = name;
        this.age = age;
        this.dob = dob;
        this.address = address;
        this.email = email;
    }
}
class UserBuilder {
    name = "";
    age = 0;
    dob = "";
    address = "";
    email = "";
    set_name(name) {
        this.name = name;
        return this;
    }
    set_age(age) {
        this.age = age;
        return this;
    }
    set_email(email) {
        this.email = email;
        return this;
    }
    set_dob(dob) {
        this.dob = dob;
        return this;
    }
    set_address(address) {
        this.address = address;
        return this;
    }
    build() {
        return new User(this.name, this.age, this.email, this.address, this.dob);
    }
}
let client = new UserBuilder().set_name("Rafi").set_age(27).set_email("rafi@evre.in");
console.log(client);

File: PythonDesignPatterns/builder.py


"""

              +----------+
              |  Client  |
              +----------+
                    |
                    | uses
                    v
          +-------------------+
          |   UserBuilder     |
          +-------------------+
          | set_name()        |
          | set_email()       |
          | set_age()         |
          | build()           |
          +-------------------+
                    |
                    | creates
                    v
              +----------+
              |   User   |
              +----------+
              | name     |
              | email    |
              | age      |
              +----------+


"""




print(__doc__)




class User:
    
    def __init__(self,name,age,email,address=None,dob=None):
        self.name=name
        self.age=age
        self.email=email
        self.address=address
        self.dob=dob


class UserBuilder:
    def __init__(self):
        self.name=None
        self.age=None
        self.email=None
        self.address=None
        self.dob=None

    def set_name(self,name):
      self.name=name
      return self
    
    def set_age(self,age):
        self.age=age
        return self
    
    def set_email(self,email):
        self.email=email
        return self

    def set_address(self,adr):
        self.address=adr
        return self
    
    def set_dob(self,dob):
        self.dob=dob
        return self
    
    def build(self):
        return User(self.name,self.age,self.email,self.address,self.dob)
    


client=UserBuilder().set_name("Raif").set_age(27).set_email("rafi@evre.in").build()
print(client)